hide_after_due for self paced courses

efischer19 opened the pull request:
# [TNL-6108](https://openedx.atlassian.net/browse/TNL-6108)

`hide_after_due` does not function in self-paced courses, as due dates are suppressed. Instead, we ought to check against the course end date in these scenarios.

efischer19 commented:
This works for hiding content after the end date of a self-paced course, as confirmed locally:
<img width="1202" alt="screen shot 2016-12-27 at 5 02 20 pm" src="https://cloud.githubusercontent.com/assets/7373924/21509366/449280e8-cc56-11e6-8c11-10b7458bf797.png">

I'll get into test updates and general cleanup from here.

efischer19 commented:
@sstack22 When you get a minute, could you confirm the basics of this feature implementation at https://efischer19.sandbox.edx.org/courses/course-v1:ericX+self+paced/courseware/9af04538e40543388fff280caff3eb09/39278b3470474368a74ce0fcb2f60c0b/? I've set up the basics of this feature request there (although doc strings haven't been touched at all).

sstack22 commented:
@efischer19 - this look awesome!! As you noted, we'll need to change the text to say 'course end date' in the advanced settings toggle, as well as the text that appears on the course outline "Subsection is hidden after due date".

I expect that some self paced courses may have already chosen the 'hide after due date' feature (thinking it would work for end date). Will those subsections automatically be hidden after this is live?

efischer19 commented:
> I expect that some self paced courses may have already chosen the 'hide after due date' feature (thinking it would work for end date). Will those subsections automatically be hidden after this is live?

They sure will! The underlying boolean is the same, I've just tweaked the "rules of hiding" to use course end date for self-paced courses.

sstack22 commented:
FYI: @catong

efischer19 commented:
@catong The sandbox is updating with my latest attempts if you'd like to take a look, particularly at the studio generalizations.

One that I'm not too sure about is the learner-facing message when encountering content that is hidden this way. What I'd like to do is essentially `"Because" + <date_type> + "has passed..."`, but since these values are translator-facing we can't build them dynamically. This is a small template anyways - we also have the option of creating another template to pair with https://github.com/edx/edx-platform/blob/master/lms/templates/hidden_content.html, specialized with the new language.

<img width="1203" alt="screen shot 2017-01-03 at 4 44 22 pm" src="https://cloud.githubusercontent.com/assets/7373924/21624279/6484c608-d1d4-11e6-93af-4ce92525976f.png">

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -12,9 +12,19 @@
+                    <%- gettext('After the course\'s end date has passed, learners can no longer access subsection content. The subsection remains included in grade calculations.') %>
+                <% } else { %>
+                    <%- gettext('After the subsection\'s due date has passed, learners can no longer access its content. The subsection remains included in grade calculations.') %>
+                <% } %>
```
I like the separate strings for the 2 cases. 
For the "course end date" case, does mentioning "remains included in grade calculations" add confusion?

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -29,7 +39,7 @@
         <p class="tip tip-warning">
             <%- interpolate(
-                gettext('If you select an option other than "%(hide_label)s", after the subsection release date has passed, published units in this subsection will become available to learners unless units are explicitly hidden.'),
+                gettext('If you select an option other than "%(hide_label)s", published units in this subsection may be made available to learners unless otherwise hidden. Please see link_from_carol_here for more information about content visibility.'),
```
Here's a suggestion: ```'If you select an option other than "%(hide_labels)s", published units in this subsection become available to learners unless they are explicitly hidden or the course has ended.'```

catong commented on lms/templates/hidden_content.html:
```diff
@@ -6,14 +6,14 @@
 <div class="sequence hidden-content proctored-exam completed">
   <h3>
-      ${_("The due date for this {subsection_format} has passed.").format(
+      ${_("The date for this {subsection_format} has passed.").format(
```
Saying "date has passed" without calling it a "due date" seems confusing.

Is it possible here to address both cases separately? (At both line 9 and line 16)
"The due date ... has passed" and "The course has ended"?

efischer19 commented on lms/templates/hidden_content.html:
```diff
@@ -5,22 +5,35 @@
+                link_end=HTML("</a>"),
+            )}
+        {% endif %}
+    </p>
```
I've standardized the tabs in this file as well, [click here to view diff ignoring whitespace changes](https://github.com/edx/edx-platform/pull/14225/files?w=1#diff-4695a24e8f2053696eabdb849e60ac0a)

efischer19 commented:
@catong I've updated the sandbox once again with your suggestions if you want to check it out.

@nasthagiri @sanfordstudent @jcdyer @yro I've squashed my work into a single commit which is ready to be reviewed at your leisure.

efischer19 commented:
jenkins run js

efischer19 commented:
jenkins run js

sstack22 commented:
@efischer19 , @catong - language looks great!

sanfordstudent commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -203,16 +203,16 @@ def handle_ajax(self, dispatch, data):  # TODO: bounds checking
+    def verify_current_content_visibility(cls, date, hide_after_date):
         """
         Returns whether the content visibility policy passes
         for the given due date and hide_after_due values and
```
docstring needs to be updated to reflect new naming.

sanfordstudent commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -286,9 +287,10 @@ def _can_user_view_content(self):
         Returns whether the runtime user can view the content
         of this sequential.
         """
+        date = self._get_course().end if getattr(self, 'self_paced', False) else self.due
```
Nit: `date` is pretty generic as far as names go.

sanfordstudent commented on common/test/acceptance/pages/lms/courseware.py:
```diff
@@ -209,13 +209,13 @@ def has_submitted_exam_message(self):
         This being true implies "the ___ contents are hidden because their due date has passed".
         """
-        message = "The due date for this {0} has passed.".format(content_type)
+        message = "this assignment is no longer available"
```
Capitalization?

efischer19 replied:
Not here - the dialog we're checking for here can vary in how it begins, so I pulled out this fragment.

<img width="675" alt="screen shot 2017-01-05 at 10 31 19 am" src="https://cloud.githubusercontent.com/assets/7373924/21686136/49f76e8a-d332-11e6-92fc-5157690be67f.png">

sanfordstudent replied:
Oooooh

sanfordstudent commented on lms/templates/hidden_content.html:
```diff
@@ -5,22 +5,35 @@
-          link_end=HTML("</a>"),
-      )}
-  </p>
+    <h3>
```
Looking at this file, I think it was the right call not to break out a separate template.

efischer19 replied:
thanks!

nasthagiri commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -12,9 +12,19 @@
-            <p class='field-message' id='hide_after_due_description'> <%- gettext('After the subsection\'s due date has passed, learners can no longer access its content. The subsection remains included in grade calculations.') %> </p>
+            <p class='field-message' id='hide_after_due_description'>
+                <% if (self_paced) { %>
+                    <%- gettext('After the course\'s end date has passed, learners can no longer access subsection content.') %>
```
Why not include "The subsection remains included in grade calculations." phrase in the self-paced case as well?
@catong @sstack22

catong replied:
I wondered in my earlier review pass whether referring to "grade calculations" might be confusing in a self-paced course - it was meant as a question not as a "let's definitely remove". If it makes sense we should retain that sentence. @sstack22 ?

nasthagiri replied:
I am suggesting to keep it for consistency with the other hide feature - otherwise, course teams may wonder whether there's a difference in grade calculations between the two hide features - when there's not.

nasthagiri commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -29,7 +39,7 @@
         <p class="tip tip-warning">
             <%- interpolate(
-                gettext('If you select an option other than "%(hide_label)s", after the subsection release date has passed, published units in this subsection will become available to learners unless units are explicitly hidden.'),
+                gettext('If you select an option other than "%(hide_label)s", published units in this subsection become available to learners unless they are explicitly hidden or the course has ended.'),
```
Doesn't this change imply that the published subsection is not available to learners after the course ends?  But that's not the case.  Even after the course ends, the learner still has read-only access to the subsection.

efischer19 replied:
Yeah, I can see that as a potential source of confusion. Would it be better just to drop that last clause and end with `If you select an option other than "%(hide_label)s", published units in this subsection become available to learners unless they are explicitly hidden`?

@catong

nasthagiri commented on lms/djangoapps/course_blocks/transformers/hidden_content.py:
```diff
@@ -38,10 +40,20 @@ def name(cls):
         return "hidden_content"
 
     @classmethod
+    def _get_merged_self_paced(cls, block_structure, block_key):
```
self-paced and course-end are properties on the course - not on each block.

So there's no reason to capture `self_paced` and "end_date" as "merged" fields, which need to be merged throughout the block hierarchy.  

Rather, simply add the following line in the `collect` method and they will be automatically collected for the course block:
```
block_structure.request_xblock_fields(u'self_paced', u'end')
```

nasthagiri replied:
Once you make this change, you should be able to get the field values by doing:

```
block_structure[block_structure.root_block_usage_key].self_paced
block_structure[block_structure.root_block_usage_key].end
```

nasthagiri commented on lms/djangoapps/course_blocks/transformers/hidden_content.py:
```diff
@@ -27,7 +27,9 @@ class HiddenContentTransformer(FilteringTransformerMixin, BlockStructureTransfor
     """
     VERSION = 1
```
You need to increment the version number on the transformer otherwise you'll get an exception in stage/production since you'll be working off of previously collected data.

efischer19 commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -97,6 +97,7 @@ def test_student_view_init(self):
     def test_render_student_view(self):
         html = self._get_rendered_student_view(
             self.sequence_3_1,
+            requested_child='first',
```
This test was failing/passing with the exact same code, I suspect it was due to a prior test requesting a different child and that child being "saved" as the current location before running this test. Does this change invalidate the assumptions under test, or is everyone okay with this?

nasthagiri replied:
Hmm.. I **do** think it invalidates the test.  As this test wants to ensure that the default position is the "first" position when a child is not explicitly requested.

The position is stored in the Courseware StudentModule SQL table - which should be getting cleared between each test run.  So I'm surprised that it's related to testing order.  

Unless, there's a cache for that data and it's not being cleared between test runs.  You can test that theory by adding the `CacheIsolationMixin` to this Test class and setting the following class variable:
```
ENABLED_CACHES = ['default', 'mongo_metadata_inheritance', 'loc_cache']
```

(Has this test ever been flaky on Jenkins since this test was introduced?)

nasthagiri commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -248,19 +248,16 @@ def _hidden_content_student_view(self, context):
                 'hidden_content.html',
                 {
-                    'subsection_format': subsection_format,
+                    'self_paced': getattr(self, 'self_paced', False),
```
I was surprised to see `self_paced` defined as an inheritable attribute on all xBlocks - and not just on the course_module.  

@rlucioni is that intentional?  Shouldn't `self_paced` be defined on just the course and not on each xBlock?  It's currently in both of these locations, which seems like an error:

- https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/course_module.py#L784
- https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/inheritance.py#L214

rlucioni replied:
@nasthagiri this is intentional, it was introduced in https://github.com/edx/edx-platform/pull/11805 to resolve [ECOM-3733](https://openedx.atlassian.net/browse/ECOM-3733), an issue preventing students in self-paced courses from seeing all available discussion modules.

nasthagiri replied:
@rlucioni Thanks for pulling up the PR where this was introduced.  I still don't see the rationale for making `self_paced` an inheritable field on all xBlocks - as inheritable fields are fundamentally costlier - than just querying the value from the enveloping course.
However, I did very that the [SelfPacedDateOverrideProvider](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/self_paced_overrides.py) overrides only the date fields and not the self_paced field for blocks.  So it doesn't invalidate the earlier assumption we made in this PR - where we're getting the self_paced value only from the course module.

rlucioni replied:
@nasthagiri @cpennington might remember more than me here, he and I worked on this together.

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -139,21 +140,15 @@ def test_hidden_content_before_due(self):
+    def test_hidden_content_past_due(self):
         progress_url = 'http://test_progress_link'
-        self._set_sequence_format(self.sequence_4_1, format_type)
+        self._set_sequence_format(self.sequence_4_1, None)
```
Since the message no longer uses the `format_type`, you should be able to delete the `_set_sequence_format` helper method altogether now.

nasthagiri commented on lms/djangoapps/course_blocks/transformers/hidden_content.py:
```diff
@@ -97,6 +101,10 @@ def _is_block_hidden(self, block_structure, block_key):
-        return not SequenceModule.verify_current_content_visibility(due, hide_after_due)
+        self_paced = block_structure[block_structure.root_block_usage_key].self_paced
+        if self_paced:
+            date = block_structure[block_structure.root_block_usage_key].end
```
Nit: Consider renaming this variable to `hidden_date` to be consistent with the improved naming chosen elsewhere.

nasthagiri commented on lms/templates/hidden_content.html:
```diff
@@ -5,22 +5,35 @@
+                line_break=HTML("<br>"),
+                link_start=HTML("<a href='{}'>").format(progress_url),
+                link_end=HTML("</a>"),
+            )}
```
Curious: Why choose to repeat the entire text here twice?  Rather than DRYing it and pulling out just the substring that changes (as is currently done in seq_module.py)?

efischer19 replied:
As mentioned above, combining translated fragments goes against [our i18n coding guidelines](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html) - see 12.2.1.1, item 1

nasthagiri replied:
makes sense.

efischer19 commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -111,6 +111,11 @@ def test_student_view_first_child(self):
+            self.sequence_3_1,
+            extra_context=dict(next_url='NextSequential', prev_url='PrevSequential', efischer_silly_var=True),
+        )
+        self._assert_view_at_position(html, expected_position=1)
```
These lines were copied from `test_render_student_view` above, and were used to illustrate what I mean about the test ordering. Without the above `efischer_silly_var` variable, the code from `test_render_student_view` test will *always* fail, because the code in `test_student_view_last_child` (specifically, `self._get_rendered_student_view(..., requested_child='last')`) sets a value on the sequence module that impacts later tests.

nasthagiri replied:
Looking at the code, I believe the issue is that the calls to `_set_up_module_system` and `_set_up_block` should be moved to the `setUp` method from the `setUpClass` method.  The module-system and runtime for the blocks are not getting re-initialized at the beginning of each test.

efischer19 commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -331,7 +329,7 @@ def _update_position(self, context, number_of_display_items):
         elif context.get('requested_child') == 'last':
             self.position = number_of_display_items or 1
-        elif self.position is None or self.position > number_of_display_items:
+        elif self.position is None or self.position > number_of_display_items or context.get('efischer_silly_var'):
```
This is the other half of my illustration commit. Note that the code in this block sets `self.position` to 3 as a part of `test_student_view_last_child`, and it never gets reset to the initial untouched state assumed by `test_render_student_view`, instead returning `{'position': 3}` when asked to render with no additional context.

efischer19 commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -248,19 +248,16 @@ def _hidden_content_student_view(self, context):
         """
         if not self._can_user_view_content():
-            subsection_format = (self.format or _("subsection")).lower()  # pylint: disable=no-member
+            date_reason = "course has ended" if getattr(self, 'self_paced', False) else "due date has passed"
```
Yup - see below comment for more. Sentences are the smallest blocks acceptable for translation per [our internationalization guide](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html), so we'll need to explicitly enumerate each possibility.

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -45,7 +43,7 @@ def _set_up_course_xml(cls):
         Sets up and returns XML course structure.
         """
-        course = xml.CourseFactory.build()
+        course = xml.CourseFactory.build(end=str(TOMORROW))
```
Are there any potential flakiness issues with using `TOMORROW` here?  For example, if this test runs at 23:59:59?

efischer19 replied:
I'll update with `FUTURE_DATE` and `PAST_FUTURE_DATE`, and define those to be a week or more in the future.

efischer19 commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -66,6 +66,13 @@ class SequenceFields(object):
+        help=_("ask Eric to remove this field"),
+        default=False,
+        scope=Scope.settings,
+    )
```
DEBUG: I had to add this field to be able to inspect the inherited `self_paced` value in `seq_module.py`

nasthagiri replied:
?  Since it is defined in inheritance.py, shouldn't it be available to the sequence block?

nasthagiri replied:
I believe the problem you are facing is in the way the unit tests are set up.  You shouldn't need this here.

efischer19 commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -281,14 +285,26 @@ def _gated_content_staff_banner(self):
+            current = current.get_parent()
+            if current == None:
+                raise Exception("Key assumption violated - went past root of course tree without hitting course block")
+        return current.end
```
DEBUG: this whole method was written in response to the previous code for getting the course end, `self._get_course().end`, yielding the following error in tests:

```
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 294, in _can_user_view_content
    hidden_date = self._get_course().end if getattr(self, 'self_paced', False) else self.due
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 131, in _get_course
    return self.descriptor.runtime.modulestore.get_course(self.course_id)  # pylint: disable=no-member
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1935, in __getattr__
    return getattr(self._descriptor_system, name)
AttributeError: 'InMemorySystem' object has no attribute 'modulestore'
```

efischer19 replied:
If anyone knows of a better way to get the current block's course's end date, I'm all ears. Otherwise we'll need to clean up the wording and type of the exception

nasthagiri replied:
Going up the tree to get the `course` can result in performance issues.  I believe the xBlock may have the course passed to it via the runtime.

nasthagiri replied:
There is already a `_get_course` method on `SequenceModule` that you should use instead.

efischer19 commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -139,39 +143,41 @@ def test_hidden_content_before_due(self):
+            extra_context=dict(progress_url=progress_url),
+        )
+        self.assertIn("hidden_content.html", html)
+        self.assertIn(progress_url, html)
```
This test now runs and hits the `HAHAHAHA` exception as expected.

efischer19 commented on common/lib/xmodule/xmodule/tests/xml/factories.py:
```diff
@@ -134,6 +134,13 @@ def parent(self, _create, extracted, **_):
+                for key, value in extracted._xml_node.attrib.iteritems()  # pylint: disable=no-member, protected-access
+                if key in InheritanceMixin.fields.keys()
+            }
+        )
```
I'm not 100% confident I need to be touching the code in this file, but this is what I had to do in order to get the `self_paced` value defined in `course = xml.CourseFactory.build(end=str(TOMORROW), self_paced=str(True))` to pass down through its children. Previously it was only present on the course root but no children, so the test was always assuming `False`.

nasthagiri replied:
I believe:
The inheritance of fields works through the runtime.  And since the sequence module tests you are adding to uses a mock runtime, the inheritance isn't working.

nasthagiri replied:
So rather than updating this code, we need to look into updating the sequence-module tests (at least for self-paced, if not all of them) to not use a test runtime.  Using a test runtime is causing all these issues you are facing with inheritable fields.  

Perhaps the easiest way to get a realistic runtime is to use the standard `ModulestoreTestCase` and get the sequence block from the modulestore.  This may be a violation since the sequence block is within the xmodule library and shouldn't have access to `ModulestoreTestCase` and django.  And so, you may be better off adding your self-paced related tests to a different location for integration testing, making use of django, modulestore, self-paced field overrides, etc.

efischer19 commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -248,19 +255,16 @@ def _hidden_content_student_view(self, context):
-            ).format(subsection_format=subsection_format)
+            if getattr(self, 'self_paced', False):
+                banner_text = _("Because the course has ended, this assignment is hidden from the learner.")
+                raise Exception("HAHAHAHAHA")
```
This exception will be removed, it's here as proof that I was able to have a `True` result from the call `getattr(self, 'self_paced', False)` at this point.

nasthagiri replied:
Can we instead update or add a regresssion test to prove that we hit this point?

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -3,49 +3,50 @@
+        if self_paced:
+            course = xml.CourseFactory.build(end=str(FUTURE_DATE), self_paced=str(True))
+        else:
+            course = xml.CourseFactory.build()
```
Why set the `end` date only for self_paced?  The above 4 lines may be easier to understand if they simply become:
```
course = xml.CourseFactory.build(end=str(FUTURE_DATE), self_paced=str(self.SELF_PACED))
```

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -3,49 +3,50 @@
-    @classmethod
-    def _set_up_course_xml(cls):
+    @staticmethod
+    def _set_up_course_xml(self_paced=False):
```
Nit: Since `SELF_PACED` is a class variable, no need to pass it into the method.

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -138,40 +142,42 @@ def test_hidden_content_before_due(self):
-        """
-        sequence._xmodule.format = format_type  # pylint: disable=protected-access
+
+class SelfPacedSequenceBlockTestCase(SequenceBlockTestBase):
```
Can you add a positive test to make sure the content isn't hidden before the course-end-date for self-paced courses: `test_hidden_content_before_end`?

nasthagiri commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -281,14 +285,26 @@ def _gated_content_staff_banner(self):
         Returns whether the runtime user can view the content
         of this sequential.
         """
+        hidden_date = self._get_course_end() if getattr(self, 'self_paced', False) else self.due
```
An alternative to creating tests with a real runtime to support inheritable fields, I believe you can just change the above line to:

```
course = self._get_course()
hidden_date = course.end if course.self_paced else self.due
``` 

That is, get the `self_paced` value off of the course - rather than off of the sequence module.  We have verified with Shelby that the self_paced value isn't intended to be inherited or overridden at the block level.

efischer19 commented:
@nasthagiri @sanfordstudent This is ready for another look - I've sidestepped the issues I was having with modulestore inheritance-in-tests failing by just mocking out the return values needed.

sanfordstudent commented:
Looks good but I see what look like non-random test failures.

nasthagiri commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -285,9 +282,11 @@ def _can_user_view_content(self):
         Returns whether the runtime user can view the content
         of this sequential.
         """
+        course = self._get_course()
```
Can you have the course be passed into this method, since the caller also needs to get the course?  This reduces a call to the modulestore.

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -58,38 +55,59 @@ def _set_up_course_xml(cls):
+        # The render operation will ask modulestore for the current course to get some data. As these tests were
+        # originally not written to be compatible with a real modulestore, we've mocked out the relevant return values.
+        with patch.object(SequenceModule, '_get_course') as mock_course:
+            type(mock_course.return_value).end = PropertyMock(return_value=COURSE_END_DATE)
```
Nit: Instead of mocking individual properties, why not just have `mock_course.return_value = self.course`?

nasthagiri commented on common/lib/xmodule/xmodule/tests/test_sequence.py:
```diff
@@ -138,40 +140,42 @@ def test_hidden_content_before_due(self):
+        html = self._get_rendered_student_view(
+            self.sequence_4_1,
+            extra_context=dict(progress_url=progress_url),
+            self_paced=True
```
Nitty: ending comma

efischer19 commented:
I fixed everything mentioned by @nasthagiri, and went ahead and squashed all my commits together as those changes were quite small. Will merge after Jenkins, thanks all!

efischer19 commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -199,7 +199,7 @@ class IndexQueryTestCase(ModuleStoreTestCase):
 
     @ddt.data(
-        (ModuleStoreEnum.Type.mongo, 8),
+        (ModuleStoreEnum.Type.mongo, 9),
```
Late-breaking update: query count went up by one due to fetching the course in seq_module

nasthagiri replied:
Yes, but notice that the count didn't go up for Split Modulestore.  For Split, since the sequence module is rendered within an outer "bulk-operation", the course is cached in the local thread and not refetched from Mongo.  

I'm not worried about the Old Mongo Modulestore, as we are moving away from using it.

cpennington commented:
This PR has been deployed to edx.org Staging in preparation for a release to production on Jan 13, 2017. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.

cpennington commented:
This PR has been deployed to edx.org.
